### PR TITLE
Extraction: integration with the Concordium infrastructure WIP

### PR DESCRIPTION
**Pull request by annenkov**
Remappings and wrappers for integrating our Rust extraction with the Concordium infrastructure.
Currently, produces executable code for two "counter" contracts: a simple counter and a counter with refinement types.

Extraction of the interpreter is WIP. The PR contains Rust code obtained by extracting the interpreter and modifying the result by hand. With a bit more work we should be able to generate these. I've created a special trait `DeserialAlloc` that allows for deserialisation and allocation on arena (I hope we can generalise it in the future, so it works for various allocators). I still use a mixture of deserialisers that leak a reference and `DeserialAlloc`. One fix for this would be implementing a derivation mechanism for `DeserialAlloc`

### CODE REVIEW

**Comment by spitters** [comment]
Nice!
Nit cargo,lock should be in gitignore, I understand:
https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html

**Comment by jakobbotsch** [comment]
This it is not a library it should be checked in, so it looks fine to me.

**Review by jakobbotsch — APPROVED** [review]

**Comment by jakobbotsch** [comment]
But on the other hand the "extraction/examples/concordium-extract/placeholder" file can be removed as there are other files in that subfolder.

**Comment by annenkov** [comment]
I actually pushed `Cargo.lock` accidentally, but indeed, it makes sense to have it there. It freezes the dependencies if I understand correctly.

I'll remove the placeholder file.

File: extraction/_CoqProject [hunk 1]
```diff
@@ -40,6 +40,9 @@ theories/Transform.v
 theories/TypeAnnotations.v
 theories/Utils.v
 theories/WcbvEvalAux.v
+theories/ConcordiumExtract.v
+theories/PluginExtract.v
+
```
Comment (jakobbotsch) [inline extraction/_CoqProject:45]: This file is alphabetical

File: extraction/_CoqProject [hunk 2]
```diff
@@ -57,3 +60,5 @@ examples/MidlangExtractTests.v
 examples/RustExtractTests.v
 examples/StackInterpreter.v
 examples/TypeAnnotationTests.v
+examples/RustCounter.v
+examples/RustInterp.v
```
Comment (jakobbotsch) [inline extraction/_CoqProject:64]: Ditto

**Comment by annenkov** [comment]
@jakobbotsch I've fixed the order of files on _CoqProject now

**Comment by jakobbotsch** [comment]
Feel free to merge it however you want (not sure if you want a squash or merge :) )

**Pull request merged**